crates/bevy_color/src/color_gradient.rs

```diff
@@ -0,0 +1,184 @@
+use crate::{ColorRange, Mix};
+use bevy_math::curve::{Curve, Interval};
+
+/// Represents a gradient of a minimum of 1 up to arbitrary many colors. Supported colors have to
+/// implement [`Mix`] and have to be from the same color space.
+///
+/// By default the color values are linearly interpolated.
+///
+/// This is useful for defining complex gradients or animated color transitions.
+#[derive(Debug, Clone)]
+pub struct ColorGradient<T: Mix> {
+    colors: Vec<T>,
+}
+
+impl<T: Mix> ColorGradient<T> {
+    /// Create a new [`ColorGradient`] from a collection of [mixable] types.
+    ///
+    /// This fails if there's not at least one mixable type in the collection.
+    ///
+    /// [mixable]: `Mix`
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_color::palettes::basic::*;
+    /// # use bevy_color::ColorGradient;
+    /// let gradient = ColorGradient::new([RED, GREEN, BLUE]);
+    /// assert!(gradient.is_ok());
+    /// ```
+    pub fn new(colors: impl IntoIterator<Item = T>) -> Result<Self, ColorGradientError> {
+        let colors = colors.into_iter().collect::<Vec<_>>();
+        let len = colors.len();
+        (!colors.is_empty())
+            .then(|| Self { colors })
+            .ok_or(ColorGradientError(len))
+    }
+
+    /// Converts the [`ColorGradient`] to a [`ColorCurve`] which implements the [`Curve`] trait
+    /// along with its adaptor methods
+    /// # Example
+    ///
+    /// ```
+    /// # use bevy_color::palettes::basic::*;
+    /// # use bevy_color::ColorGradient;
+    /// # use bevy_color::Srgba;
+    /// # use bevy_color::Mix;
+    /// # use bevy_math::curve::Curve;
+    /// let gradient = ColorGradient::new([RED, GREEN, BLUE]).unwrap();
+    /// let curve = gradient.to_curve();
+    ///
+    /// // you can then apply useful methods ontop of the gradient
+    /// let brighter_curve = curve.map(|c| c.mix(&WHITE, 0.25));
+    ///
+    /// assert_eq!(brighter_curve.sample_unchecked(0.0), Srgba::new(1.0, 0.25, 0.25, 1.0));
+    /// ```
+    pub fn to_curve(self) -> ColorCurve<T> {
+        let domain =
+            Interval::new(0.0, (self.colors.len() - 1) as f32).expect("at least 1 by construction");
```

Comment:
This contradicts the documentation for `ColorRange::at`, which implies that a `ColorRange<T>` is basically a curve parametrized by the unit interval.